spark/src/main/java/org/apache/iceberg/spark/source/StreamingReader.java

```diff
@@ -0,0 +1,292 @@
+import org.apache.iceberg.Snapshot;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.TableProperties;
+import org.apache.iceberg.encryption.EncryptionManager;
+import org.apache.iceberg.io.CloseableIterable;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.util.SnapshotUtil;
+import org.apache.iceberg.util.TableScanUtil;
+import org.apache.spark.broadcast.Broadcast;
+import org.apache.spark.sql.sources.v2.DataSourceOptions;
+import org.apache.spark.sql.sources.v2.reader.streaming.MicroBatchReader;
+import org.apache.spark.sql.sources.v2.reader.streaming.Offset;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A mirco-batch based Spark Structured Streaming reader for Iceberg table. It will track the added
+ * files and generate tasks per batch to process newly added files. By default it will process
+ * all the newly added files to the current snapshot in each batch, user could also set this
+ * configuration "max-files-per-trigger" to control the number of files processed per batch.
+ */
+class StreamingReader extends Reader implements MicroBatchReader {
+  private static final Logger LOG = LoggerFactory.getLogger(StreamingReader.class);
+  private static final int DEFAULT_MAX_FILES_PER_TRIGGER = 1000;
+  private static final long CACHE_SIZE = 1000;
+
+  private Offset startOffset;
+  private Offset endOffset;
+
+  private final Table table;
+  private final int maxFilesPerTrigger;
+  private final long splitSize;
+  private final int lookback;
+  private final long openFileCost;
+  private final Offset initialOffset;
+
+  private final Cache<Long, CloseableIterable<FileScanTask>> snapshotTasksCache = Caffeine.newBuilder()
+      .maximumSize(CACHE_SIZE)
+      .softValues()
+      .build();
+
+  /**
+   * Utility class to track the snapshotId, task and index of task within this snapshot.
+   */
+  @VisibleForTesting
+  static class IndexedTask {
+    private final long snapshotId;
+    private final FileScanTask task;
+    private final int index;
+
+    IndexedTask(long snapshotId, FileScanTask task, int index) {
+      this.snapshotId = snapshotId;
+      this.task = task;
+      this.index = index;
+    }
+
+    long snapshotId() {
+      return snapshotId;
+    }
+
+    FileScanTask task() {
+      return task;
+    }
+
+    int index() {
+      return index;
+    }
+  }
+
+  StreamingReader(Table table, Broadcast<FileIO> io, Broadcast<EncryptionManager> encryptionManager,
+                  boolean caseSensitive, DataSourceOptions options) {
+    super(table, io, encryptionManager, caseSensitive, options);
+
+    this.table = table;
+    this.maxFilesPerTrigger =
+      options.get("max-files-per-trigger").map(Integer::parseInt).orElse(DEFAULT_MAX_FILES_PER_TRIGGER);
+    Preconditions.checkArgument(maxFilesPerTrigger > 0,
+        "Option max-files-per-trigger '%d' should > 0", maxFilesPerTrigger);
+
+    Long snapshotId = options.get("snapshot-id").map(Long::parseLong).orElse(null);
```

Comment:
Should this be `starting-snapshot-id`?

Also, is this called when resuming from a checkpoint or is this only when creating a completely new reader?

Reply:
This will be leveraged only when creating a completely new reader. If it is resuming from a checkpoint, Spark will pass in the checkpointed `startOffset` in `setOffsetRange()`.